src/animation/gravity.ts

```diff
@@ -0,0 +1,119 @@
+import { vector as Vector, subf, magf } from "../core/Vector.gen";
+import { Entity, applyForce } from "../forces/Force.gen";
+import { forceV } from "../forces/Gravity.gen";
+import { rAF } from "../rAF";
+
+export interface GravityParams {
+  config: {
+    moverMass: number;
+    attractorMass: number;
+    attractor: Vector<number>;
+    initialVelocity?: Vector<number>;
+    threshold?: {
+      min: number;
+      max: number;
+    };
+  };
+  onFrame: (position: Vector<number>) => void;
```

Comment:
I'm wondering if `onFrame` should have velocity in the params. I think this could be useful for the user in several cases, but off the top of my head:

- Would allow for them to switch between forces, by preserving velocity between the swap
- Might be nice for animating certain properties. I know squash n stretch (scale) can be a common velocity based property.

Reply:
Just a thought here too, and this is super minor, but do you think the concept `frames` is an internal concept that doesn't need to be exposed in the outward facing api?

maybe having an `onPositionChange` & `onVelocityChange` or `onUpdate` might seem a little friendlier to the user.

Just a thought tho. I think you can make an argument for what you have here as well that it's more transparent as to what's going on under the hood

Reply:
These are all good shouts! I could definitely see the API being:

```typescript
onUpdate: ({ position, velocity }) => {
   // Mutate your DOM how you like with either property.
}
```

It still feels a bit early to say though. The question I'm really wrestling with is figuring out the bounds of what we do vs. what the user does. I feel like Popmotion strikes a nice balance – sane defaults with options for configuration and customization. `react-spring` feels more opinionated, where it just handles it all for you and you pass these opaque props onto its opaque rendering target. Ultimately, `onUpdate` may be kept internal to `useGravity` if we force a rendering target on users (i.e. `animated.*` elements) or it may be opened up if we don't!

Reply:
yea real good point. I'm excited to see where you take it once you get more into react land